klippy/extras/tmc2660.py

```diff
@@ -218,52 +220,27 @@ def build_config(self):
             "spi_transfer oid=%c data=%*s", cq=cmd_queue)
 
-    # register timeout handler which will lower the current to current * idle_current_percent / 100 after idle_timeout seconds
-    # and raise it back to current if the printer needs to move the steppers
-    def printer_state(self, state):
-        if state == 'ready' and self.idle_timeout > 0:
-            self.toolhead = self.printer.lookup_object('toolhead')
-            reactor = self.printer.get_reactor()
-            reactor.register_timer(self.idle_timeout_handler, reactor.NOW)
+    def handle_printing(self, print_time):
+        self.set_current(print_time, self.current)
 
-    # timeout handler to lower/raise current when entering/leaving the idle state
-    def idle_timeout_handler(self, eventtime):
-        info = self.toolhead.get_status(eventtime)
-        status = info['status']
-        print_time = info['print_time']
-        if status == 'Printing':
-            if self.is_idle:
-                self.set_current(self.current)
-                self.is_idle = False
-            return eventtime + self.idle_timeout
-        estimated_print_time = info['estimated_print_time']
-        elapsed_time = estimated_print_time - print_time
-        if elapsed_time < self.idle_timeout:
-            if self.is_idle:
-                self.set_current(self.current)
-                self.is_idle = False
-            return eventtime + self.idle_timeout - elapsed_time
-        if not self.is_idle:
-            self.set_current(float(self.idle_current_percentage) * self.current / 100)
-            self.is_idle = True
-        return eventtime + 0.1
+    def handle_ready(self, print_time):
+        self.set_current(print_time, float(self.idle_current_percentage) * self.current / 100)
 
-    def set_current(self, current):
+    def set_current(self, print_time, current):
         self.driver_cs = current_to_reg(current)
         reg = self.reg_sgcsconf
         reg &= ~(SGCSCONF["CS"][1])
         reg |= get_bits(SGCSCONF, "CS", self.driver_cs)
         reg_data = [(reg >> 16) & 0xff, (reg >> 8) & 0xff, reg & 0xff]
-        params = self.spi_transfer_cmd.send_with_response([self.oid, reg_data], 'spi_transfer_response', self.oid)
+        clock = self.mcu.print_time_to_clock(print_time)
+        params = self.spi_send_cmd.send([self.oid, reg_data], minclock=clock, reqclock=clock)
 
     cmd_SET_TMC_CURRENT_help = "Set the current of a TMC2660 driver (between %d and %d)" % (CURRENT_MIN, CURRENT_MAX)
     def cmd_SET_TMC_CURRENT(self, params):
-        self.printer.lookup_object('toolhead').get_last_move_time()
         gcode = self.printer.lookup_object('gcode')
         if 'CURRENT' in params:
             self.current = gcode.get_float('CURRENT', params, minval=CURRENT_MIN, maxval=CURRENT_MAX)
-            if not self.is_idle:
-                self.set_current(self.current)
+            self.set_current(self.current, self.printer.get_reactor().NOW)
```

Comment:
This line doesn't look right - set_current() takes print_time followed by a current..

Also, the reactor works in system times, and they shouldn't be mixed with print_times - the toolhead.get_last_move_time() method returns a print_time that could be used here.

Reply:
oh dear, good catch! I'll fixc it right away!